### DISCUSSION TITLE
How to prevent 'update cascade' when using the trigger 'items.update (action)' ?

### DISCUSSION THREAD

**Discussion by arladmin** [directus/directus#16181] [Archive]
I've implemented a Flow with Trigger as item.update (action) which then passes on the trigger data to my external server.

My server then does some operations, modifies the input values and then updates the trigger record with the new values (via the REST api).

The issue now is, that this triggers an endless update cascade: 
**Update via Directus app -> Action trigger fired -> item updated via REST api -> Action trigger fired -> .....**

I'm sure i'm missing something very obvious here.

Would be grateful if anyone can point that out and help me in resolving this issue.

Thanks

**Comment by br41nslug** [comment]
To prevent such an endless loop, if step `item updated via REST api` is done using an "Update Data" operation that is as the API does not have this feature, you can disable the firing of these events by toggling "emit events" off (which should be the default):

![image](https://user-images.githubusercontent.com/9389634/198031314-439a6b72-d46e-4d16-a0c0-4eb04a2bc8d3.png)

**Reply by arladmin** [reply]
@br41nslug Thanks for the response!

I did think of using the 'Update data" operation, but, i'm currently executing a whole bunch of operations on my server, via the REST api, and would really prefer not to have to transfer all of that logic to a Directus Flow.

Hence, looking for a solution for the REST api itself.

**Reply by br41nslug** [reply]
> i'm currently executing a whole bunch of operations on my server

How do you mean a bunch of operations? is this not a directus enpoint you're calling or are you doing a lot of requests?

> and would really prefer not to have to transfer all of that logic to a Directus Flow.

Understandably however there are some arguments for the create/read/update operations as they're talking directly to the database instead of going over the internet to send data to itself.

> Hence, looking for a solution for the REST api itself.

As i mentioned before this diabling events not a feature in the API but is available for Flows / API extensions. If there is no other option than doing API call(s) you'll need to solve this at the "trigger" level (most likely replace the flow with a custom hook)

**Reply by arladmin** [reply]
So, the flow is-
**Trigger (Event Hook - action/items.update)  ->  Request URL (body: `$trigger` )**
<hr>

> How do you mean a bunch of operations? is this not a directus enpoint you're calling or are you doing a lot of requests?
> 
> >
Then, when i receive the HTTP call, i use the trigger data and update a bunch of records in various collections based on that, via the REST api. This processes involves quite a few conditional read/update queries for multiple collections, which would be really cumbersome to implement via a Directus Flow.
<hr>

I was hoping to be able to find a solution for this with the REST api itself, but like you suggest, it may currently not be possible.

**Comment by br41nslug** [comment]
> Then, when i receive the HTTP call, i use the trigger data and update a bunch of records in various collections based on that, via the REST api. This processes involves quite a few conditional read/update queries for multiple collections, which would be really cumbersome to implement via a Directus Flow.

As a compromise you could consider implementing this logic in a custom endpoint (which has access to the `ItemsServices` on the server that have this "prevent emitting events" feature too) and call that from the flow or you could even do it all in a custom hook instead of endpoint replacing the flow too.

**Reply by arladmin** [reply]
Yeah, that's a good suggestion.

Although, the main requirement in this use case is to be able to do all of this externally.

For now, will probably just use Flows or Custom Hooks.

Will also raise a Feature Request to implement the 'event emitter disable option' within the REST api as that seems the only sustainable method for such use-cases.

**Reply by br41nslug** [reply]
Hmmm the "be able to do all of this externally." does make it tricky 🤔 It's a bit more hacky but a workaround you could do is add an extra field on the collection (let's say `events` any type should do) and fill that field when doing the external API calls then checking whether that field exists in the update payload in the flow. If you find the field you can throw an error (using the Script operation) which will abort the flow. You did say `action` trigger right? this won't work for `filter`.

**Reply by arladmin** [reply]
That's a crafty solution! Thanks.
Have created a feature request #16182 though, for a more permanent resolution.
